File: Sources/XMTP/Messages/MessageV1.swift
//
//  MessageV1.swift
//
//
//  Created by Pat Nakajima on 11/26/22.
//

import Foundation
import XMTPProto

typealias MessageV1 = Xmtp_MessageContents_MessageV1

enum MessageV1Error: Error {
	case cannotDecodeFromBytes
}

extension MessageV1 {
	static func encode(sender: PrivateKeyBundleV1, recipient: PublicKeyBundle, message: Data, timestamp: Date) throws -> MessageV1 {
		let secret = try sender.sharedSecret(
			peer: recipient,
			myPreKey: sender.preKeys[0].publicKey,
			isRecipient: false
		)

		let header = MessageHeaderV1(
			sender: sender.toPublicKeyBundle(),
			recipient: recipient,
			timestamp: UInt64(timestamp.millisecondsSinceEpoch)
		)

		let headerBytes = try header.serializedData()
		let ciphertext = try Crypto.encrypt(secret, message, additionalData: headerBytes)

		return MessageV1(headerBytes: headerBytes, ciphertext: ciphertext)
	}

	static func fromBytes(_ bytes: Data) throws -> MessageV1 {
		let message = try Message(serializedData: bytes)
		var headerBytes: Data
		var ciphertext: CipherText

		switch message.version {
		case .v1:
			headerBytes = message.v1.headerBytes
			ciphertext = message.v1.ciphertext
		case .v2:
			headerBytes = message.v2.headerBytes
			ciphertext = message.v2.ciphertext
		default:
			throw MessageV1Error.cannotDecodeFromBytes
		}

		return MessageV1(headerBytes: headerBytes, ciphertext: ciphertext)
	}

	init(headerBytes: Data, ciphertext: CipherText) {
		self.init()
		self.headerBytes = headerBytes
		self.ciphertext = ciphertext
	}

	var header: MessageHeaderV1 {
		get throws {
			do {
				return try MessageHeaderV1(serializedData: headerBytes)
			} catch {
				print("Error deserializing MessageHeaderV1 \(error)")
				throw error
			}
		}
	}

	var senderAddress: String? {
		do {
			let senderKey = try header.sender.identityKey.recoverWalletSignerPublicKey()
			return senderKey.walletAddress
		} catch {
			print("Error getting sender address: \(error)")
			return nil
		}
	}

	var sentAt: Date {
		// swiftlint:disable force_try
		try! Date(timeIntervalSince1970: Double(header.timestamp / 1000))
		// swiftlint:enable force_try
	}

	var recipientAddress: String? {
		do {
			let recipientKey = try header.recipient.identityKey.recoverWalletSignerPublicKey()

			return recipientKey.walletAddress
		} catch {
			print("Error getting recipient address: \(error)")
			return nil
		}
	}

	func decrypt(with viewer: PrivateKeyBundleV1) throws -> Data {
		let header = try MessageHeaderV1(serializedData: headerBytes)

		let recipient = header.recipient
		let sender = header.sender

		var secret: Data
		if viewer.walletAddress == sender.walletAddress {
			secret = try viewer.sharedSecret(peer: recipient, myPreKey: sender.preKey, isRecipient: false)
		} else {
			secret = try viewer.sharedSecret(peer: sender, myPreKey: recipient.preKey, isRecipient: true)
		}

		return try Crypto.decrypt(secret, ciphertext, additionalData: headerBytes)
	}
}

File: library/src/main/java/org/xmtp/android/library/messages/MessageV1.kt
package org.xmtp.android.library.messages

import com.google.protobuf.kotlin.toByteString
import org.xmtp.android.library.CipherText
import org.xmtp.android.library.Crypto
import org.xmtp.proto.message.contents.MessageOuterClass
import java.util.Date

typealias MessageV1 = org.xmtp.proto.message.contents.MessageOuterClass.MessageV1

class MessageV1Builder {
    companion object {
        fun buildEncode(
            sender: PrivateKeyBundleV1,
            recipient: PublicKeyBundle,
            message: ByteArray,
            timestamp: Date
        ): MessageV1 {
            val secret = sender.sharedSecret(
                peer = recipient,
                myPreKey = sender.preKeysList[0].publicKey,
                isRecipient = false
            )
            val header = MessageHeaderV1Builder.buildFromPublicBundles(
                sender = sender.toPublicKeyBundle(),
                recipient = recipient,
                timestamp = timestamp.time
            )
            val headerBytes = header.toByteArray()
            val ciphertext = Crypto.encrypt(secret, message, additionalData = headerBytes)
            return buildFromCipherText(headerBytes = headerBytes, ciphertext = ciphertext)
        }

        fun buildFromBytes(bytes: ByteArray): MessageV1 {
            val message = Message.parseFrom(bytes)
            val headerBytes: ByteArray
            val ciphertext: CipherText
            when (message.versionCase) {
                MessageOuterClass.Message.VersionCase.V1 -> {
                    headerBytes = message.v1.headerBytes.toByteArray()
                    ciphertext = message.v1.ciphertext
                }
                MessageOuterClass.Message.VersionCase.V2 -> {
                    headerBytes = message.v2.headerBytes.toByteArray()
                    ciphertext = message.v2.ciphertext
                }
                else -> throw IllegalArgumentException("Cannot decode from bytes")
            }
            return buildFromCipherText(headerBytes, ciphertext)
        }

        fun buildFromCipherText(headerBytes: ByteArray, ciphertext: CipherText?): MessageV1 {
            return MessageV1.newBuilder().also {
                it.headerBytes = headerBytes.toByteString()
                it.ciphertext = ciphertext
            }.build()
        }
    }
}

val MessageV1.header: MessageHeaderV1
    get() = MessageHeaderV1.parseFrom(headerBytes)

val MessageV1.senderAddress: String
    get() = header.sender.identityKey.recoverWalletSignerPublicKey().walletAddress

val MessageV1.sentAt: Date get() = Date(header.timestamp)

val MessageV1.recipientAddress: String
    get() = header.recipient.identityKey.recoverWalletSignerPublicKey().walletAddress

fun MessageV1.decrypt(viewer: PrivateKeyBundleV1?): ByteArray? {
    val header = MessageHeaderV1.parseFrom(headerBytes)
    val recipient = header.recipient
    val sender = header.sender
    val secret: ByteArray = if (viewer?.walletAddress == sender.walletAddress) {
        viewer.sharedSecret(peer = recipient, myPreKey = sender.preKey, isRecipient = false)
    } else {
        viewer?.sharedSecret(peer = sender, myPreKey = recipient.preKey, isRecipient = true)
            ?: byteArrayOf()
    }
    return Crypto.decrypt(secret, ciphertext, additionalData = headerBytes.toByteArray())
}
